arraycontext/impl/pytato/__init__.py

```diff
@@ -136,13 +139,30 @@ def _get_fake_numpy_namespace(self):
         from arraycontext.impl.pytato.fake_numpy import PytatoFakeNumpyNamespace
         return PytatoFakeNumpyNamespace(self)
 
+    @abc.abstractproperty
+    def frozen_array_types(self) -> Tuple[Type, ...]:
+        """
+        Returns valid frozen array types for the array context.
+        """
+
+    # {{{ ArrayContext interface
+
     def empty(self, shape, dtype):
-        raise ValueError(f"{type(self).__name__} does not support empty")
+        raise NotImplementedError(
+            f"{type(self).__name__}.empty is not supported")
```

Comment:
I think we should deprecate `ArrayContext.empty` and `empty_like`.

Reply:
At least `pytential` seems to use that quite a bit to preallocate some arrays.

I'll add the deprecation in a follow-up PR, so we can talk about it a bit more.

Reply:
Sounds good!